src/cmd/tools/dtest/docker/harness/query_api_test.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package harness
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/m3db/m3/src/cmd/tools/dtest/docker/harness/resources"
+)
+
+var instantQueryBadRequestTest = []struct {
+	name string
+	q    string
+}{
+	// {"missing query", queryURL("query", "")},
+	// FAILING issue #2: invalid query string should result in 400
+	// {"invalid query", queryURL("query", "@!")},
+	{"invalid time", queryURL("time", "INVALID")},
+	{"invalid timeout", queryURL("timeout", "INVALID")},
+}
+
+func TestInvalidInstantQueryReturns400(t *testing.T) {
+	coord := singleDBNodeDockerResources.Coordinator()
+
+	for _, tt := range instantQueryBadRequestTest {
+		t.Run(tt.name, func(t *testing.T) {
+			assert.NoError(t, coord.RunQuery(verifyStatus(400), tt.q), "for query '%v'", tt.q)
+		})
+	}
+}
+
+var queryBadRequestTest = []struct {
+	name string
+	q    string
```

Comment:
Nit: this struct is declared twice, better have it as one explicit struct.

Reply:
https://github.com/m3db/m3/pull/2873/commits/77f076c2efb862b562c7c349e6247d21a2ba78d3#diff-391a291f1f91098bc731847f8843207e5327fe87a52a4599a45ac0cf4c5732e7R33